File: frontend/src/types/scanner.ts
/**
 * Scanner Control Types (Story 20.6)
 *
 * TypeScript interfaces for scanner control UI matching backend models.
 */

/**
 * Valid timeframes for scanner watchlist
 */
export type ScannerTimeframe =
  | '1M'
  | '5M'
  | '15M'
  | '30M'
  | '1H'
  | '4H'
  | '1D'
  | '1W'

/**
 * Valid asset classes for scanner watchlist
 */
export type ScannerAssetClass = 'forex' | 'stock' | 'index' | 'crypto'

/**
 * Scanner state values
 */
export type ScannerState =
  | 'stopped'
  | 'starting'
  | 'running'
  | 'waiting'
  | 'scanning'
  | 'stopping'

/**
 * Response from GET /api/v1/scanner/status
 */
export interface ScannerControlStatus {
  is_running: boolean
  current_state: ScannerState
  last_cycle_at: string | null
  next_scan_in_seconds: number | null
  symbols_count: number
  scan_interval_seconds: number
  session_filter_enabled: boolean
}

/**
 * Response from POST /api/v1/scanner/start or /stop
 */
export interface ScannerActionResponse {
  status: 'started' | 'stopped' | 'already_running' | 'already_stopped'
  message: string
  is_running: boolean
}

/**
 * Symbol entry in scanner watchlist
 */
export interface ScannerWatchlistSymbol {
  id: string
  symbol: string
  timeframe: ScannerTimeframe
  asset_class: ScannerAssetClass
  enabled: boolean
  last_scanned_at: string | null
  created_at: string
  updated_at: string
}

/**
 * Request body for adding a symbol to scanner watchlist
 */
export interface AddScannerSymbolRequest {
  symbol: string
  timeframe: ScannerTimeframe
  asset_class: ScannerAssetClass
}

/**
 * Request body for updating a symbol in scanner watchlist
 */
export interface UpdateScannerSymbolRequest {
  enabled: boolean
}

/**
 * Scan cycle history record
 */
export interface ScannerHistoryRecord {
  id: string
  cycle_started_at: string
  cycle_ended_at: string | null
  symbols_scanned: number
  signals_generated: number
  errors_count: number
  status: 'COMPLETED' | 'PARTIAL' | 'FAILED' | 'SKIPPED' | 'FILTERED'
}

/**
 * WebSocket scanner status changed event
 */
export interface ScannerStatusChangedEvent {
  type: 'scanner:status_changed'
  is_running: boolean
  event: 'started' | 'stopped' | 'cycle_completed'
  timestamp: string
  sequence_number: number
}

/**
 * Timeframe options for dropdowns
 */
export const TIMEFRAME_OPTIONS: { label: string; value: ScannerTimeframe }[] = [
  { label: '1 Minute', value: '1M' },
  { label: '5 Minutes', value: '5M' },
  { label: '15 Minutes', value: '15M' },
  { label: '30 Minutes', value: '30M' },
  { label: '1 Hour', value: '1H' },
  { label: '4 Hours', value: '4H' },
  { label: '1 Day', value: '1D' },
  { label: '1 Week', value: '1W' },
]

/**
 * Asset class options for dropdowns
 */
export const ASSET_CLASS_OPTIONS: {
  label: string
  value: ScannerAssetClass
}[] = [
  { label: 'Forex', value: 'forex' },
  { label: 'Stock', value: 'stock' },
  { label: 'Index', value: 'index' },
  { label: 'Crypto', value: 'crypto' },
]

/**
 * Maximum watchlist size
 */
export const MAX_WATCHLIST_SIZE = 50

File: backend/src/orm/scanner.py
"""
SQLAlchemy ORM Models for Scanner Persistence (Story 20.1).

Models for scanner state persistence including:
- ScannerWatchlistORM: Symbols to scan
- ScannerConfigORM: Scanner configuration (singleton)
- ScannerHistoryORM: Scan cycle history

These tables persist scanner state across service restarts.
"""

from datetime import UTC, datetime
from uuid import UUID, uuid4

from sqlalchemy import JSON, Boolean, CheckConstraint, Index, Integer, String
from sqlalchemy.dialects.postgresql import JSONB as PG_JSONB
from sqlalchemy.dialects.postgresql import TIMESTAMP
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import TypeDecorator


class PortableJSONB(TypeDecorator):
    """JSONB on PostgreSQL, JSON on other dialects (e.g. SQLite in tests)."""

    impl = JSON
    cache_ok = True

    def load_dialect_impl(self, dialect):
        if dialect.name == "postgresql":
            return dialect.type_descriptor(PG_JSONB())
        return dialect.type_descriptor(JSON())


from src.database import Base


class ScannerWatchlistORM(Base):
    """
    Scanner watchlist entry for persistent symbol tracking.

    Table: scanner_watchlist
    Primary Key: id (UUID)
    Unique: symbol

    Stores symbols that the scanner should process, along with
    their timeframe and asset class configuration.
    """

    __tablename__ = "scanner_watchlist"

    # Primary key
    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    # Symbol identification
    symbol: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        unique=True,
    )

    # Timeframe: 1M, 5M, 15M, 30M, 1H, 4H, 1D, 1W
    timeframe: Mapped[str] = mapped_column(
        String(10),
        nullable=False,
    )

    # Asset class: forex, stock, index, crypto
    asset_class: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    # Whether this symbol is actively scanned
    enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        server_default="true",
    )

    # Last time this symbol was scanned
    last_scanned_at: Mapped[datetime | None] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=True,
    )

    # When this symbol was added to watchlist
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True),
        default=lambda: datetime.now(UTC),
        nullable=False,
    )

    # When this symbol was last modified
    updated_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
        nullable=False,
    )

    # Note: PostgreSQL-specific constraints like symbol ~ '^[A-Z0-9./^-]+$'
    # are defined in the migration. The ORM constraints are SQLite-compatible
    # for testing purposes.
    __table_args__ = (
        CheckConstraint(
            "LENGTH(symbol) BETWEEN 1 AND 20",
            name="chk_scanner_watchlist_symbol_length",
        ),
        CheckConstraint(
            "timeframe IN ('1M', '5M', '15M', '30M', '1H', '4H', '1D', '1W')",
            name="chk_scanner_watchlist_timeframe",
        ),
        CheckConstraint(
            "asset_class IN ('forex', 'stock', 'index', 'crypto')",
            name="chk_scanner_watchlist_asset_class",
        ),
        Index(
            "idx_scanner_watchlist_enabled",
            "enabled",
        ),
    )


class ScannerConfigORM(Base):
    """
    Scanner configuration (singleton pattern).

    Table: scanner_config
    Primary Key: id (UUID)

    Stores scanner settings like scan interval, batch size,
    session filtering, and running state. Only one row exists.
    """

    __tablename__ = "scanner_config"

    # Primary key
    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    # How often to scan (in seconds), minimum 60
    scan_interval_seconds: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        server_default="300",
    )

    # How many symbols to process per batch (1-50)
    batch_size: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        server_default="10",
    )

    # Whether to filter by trading session hours
    session_filter_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        server_default="true",
    )

    # Whether scanner is currently running
    is_running: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        server_default="false",
    )

    # When last scan cycle completed
    last_cycle_at: Mapped[datetime | None] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=True,
    )

    # When config was last updated
    updated_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "scan_interval_seconds >= 60",
            name="chk_scanner_config_interval",
        ),
        CheckConstraint(
            "batch_size >= 1 AND batch_size <= 50",
            name="chk_scanner_config_batch_size",
        ),
    )


class ScannerHistoryORM(Base):
    """
    Scanner history for tracking scan cycles.

    Table: scanner_history
    Primary Key: id (UUID)
    Indexes: idx_scanner_history_cycle_started

    Records each scan cycle's results for monitoring
    and debugging purposes.
    """

    __tablename__ = "scanner_history"

    # Primary key
    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    # When this scan cycle started
    cycle_started_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
    )

    # When this scan cycle ended (null if still running or crashed)
    cycle_ended_at: Mapped[datetime | None] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=True,
    )

    # Number of symbols scanned in this cycle
    symbols_scanned: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    # Number of signals generated in this cycle
    signals_generated: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    # Number of errors encountered
    errors_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        server_default="0",
    )

    # Number of symbols with no OHLCV data
    symbols_no_data: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        server_default="0",
    )

    # Correlation IDs of signals generated during this cycle (Task #25)
    # Stored as JSONB array of UUID strings: ["uuid1", "uuid2", ...]
    correlation_ids: Mapped[list[str] | None] = mapped_column(
        PortableJSONB,
        nullable=True,
        server_default="[]",
    )

    # Cycle status: COMPLETED, PARTIAL, FAILED, SKIPPED
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "status IN ('COMPLETED', 'PARTIAL', 'FAILED', 'SKIPPED')",
            name="chk_scanner_history_status",
        ),
        Index(
            "idx_scanner_history_cycle_started",
            "cycle_started_at",
        ),
    )
